lib/policies/redirectPolicy.ts

```diff
@@ -44,23 +44,34 @@ function handleRedirect(
   const locationHeader = response.headers.get("location");
   if (
     locationHeader &&
-    (status === 300 || status === 307 || (status === 303 && request.method === "POST")) &&
-    (!policy.maxRetries || currentRetries < policy.maxRetries)
-  ) {
+    (status === 300 || status === 302 || status === 307 || (status === 303 && request.method === "POST")) &&
+    (request.redirectLimit && currentRetries < request.redirectLimit)) {
     const builder = URLBuilder.parse(request.url);
     builder.setPath(locationHeader);
     request.url = builder.toString();
 
-    // POST request with Status code 303 should be converted into a
+    // POST request with Status code 302 and 303 should be converted into a
     // redirected GET request if the redirect url is present in the location header
-    if (status === 303) {
+    // reference: https://tools.ietf.org/html/rfc7231#page-57 && https://fetch.spec.whatwg.org/#http-redirect-fetch
+    if (status === 302 || status === 303) {
```

Comment:
Reading through the specs confuses me a little; the IETF link says for 302:

```
Note: For historical reasons, a user agent MAY change the request
      method from POST to GET for the subsequent request.
```

Which doesn't seem to motivate strong action. Meanwhile, the fetch spec notes we should switch the method to GET if:

```
actualResponse’s status is 301 or 302 and request’s method is `POST`
actualResponse’s status is 303 and request’s method is not `GET` or `HEAD`
```

But we're not preserving `HEAD` in the 303 case, nor are we handling 301 at all.

Reply:
Yes I found the `for historical reasons` comment not very helpful. I would recommend following the node-fetch implementation as that is what is currently being applied and will cause the minimum disruption to users of this library.

I was initially tempted to port [Azure/azure-sdk-for-js#11863](https://github.com/Azure/azure-sdk-for-js/pull/11863) into this library but choose not to as the behaviour was different from the [node-fetch spec](https://fetch.spec.whatwg.org/#http-redirect-fetch). Specifically regarding the 302 POST redirected as a GET which occurs in one of the SAML flows we are testing.

So where does that leave us with this PR. I would suggest either: 
 1. Respect max retires (even though we know it is not currently used)
 2. Port the `redirectPolicy` changes from `azure-sdk-for-js`, extend the redirect behaviour in line with the `node-fetch` spec and then add the `request.redirectLimit`. 

I'll do option 1 now and see if I have time for option 2.
Any thoughts?

Reply:
I have a preference for the second option if you're able to, just because keeping this implementation closer to the one in `core-http` will yield easier maintenance in the future.

If the 302 POST becoming a GET is important, I'm fine with adding that behavior, though I think your if check at line 47 should be modified to have

```
(status === 302 && request.method === "POST)"
```

instead of simply `status === 302` since we don't want to convert non-POST 302s to GETs, right?

Reply:
I agree, option 2 seems the correct thing to do. I'm in no rush to merge this so I'll have a look at implementing option 2 when I get a chance.